### PR TITLE
refactor: plumbing to prevent compaction starving persistence

### PR DESCRIPTION
**Pull request by tustvold**
Once persistence needs to operate on potentially multiple chunks it creates a situation where one of the criteria to persist has been met, but it is unable proceed because of in-progress compactions on one or more of the source chunks.

This PR therefore does two things:

* It pulls the partition loop out of `maybe_*` into a single loop
* Adds a boolean return that `maybe_persist_chunks` can use to indicate that compaction should be stalled

This currently isn't a problem as persistence currently only operates on a single chunk, and the decision to persist a chunk takes place before the decision to compact it

### CODE REVIEW

File: lifecycle/src/policy.rs [hunk 1]
```diff
@@ -1320,10 +1324,10 @@ mod tests {
         assert_eq!(
             *db.events.read(),
             vec![
+                MoverEvents::Compact(vec![0, 1]),
```
Comment (tustvold) [inline lifecycle/src/policy.rs:1327]: Pulling the loop iterations out causes this to be reordered

File: lifecycle/src/policy.rs [hunk 2]
```diff
@@ -247,76 +244,74 @@ where
...
+                    .expect("task preparation failed")
+                    .with_metadata(ChunkLifecycleAction::Persisting);
 
-                    self.trackers.push(tracker);
-                }
+                self.trackers.push(tracker);
             }
         }
+
+        false
```
Comment (crepererum) [inline lifecycle/src/policy.rs:314]: Is `false` the only possible return value for now? There don't seem to be any other return (or I'm a bit blind).

Comment (tustvold) [reply lifecycle/src/policy.rs:314]: Correct - this PR is just putting the plumbing in place, but not using it

**Review by alamb — APPROVED** [review]
makes sense to me -- and is a nice independent refactoring in advance of persistence

**Pull request merged**